### DISCUSSION TITLE
Jupyter error "tried to read a stream that is not readable"

### DISCUSSION THREAD

**Discussion by quasars100** [MilesCranmer/PySR#248] [Q&A] [answered]
Hi,
Sorry this is a basic question about installation - I followed all instructions (Julia was already installed, then I ran "pip install -U pysr" followed by " python -c 'import pysr; pysr.install()' ").  There was no issues regarding importing Pysr in python; however, when I try running
```py
model = PySRRegressor(
    niterations=40,  # < Increase me for better results
    binary_operators=["+", "*"],
    unary_operators=[
        "cos",
        "exp",
        "sin",
        "inv(x) = 1/x",
        # ^ Custom operator (julia syntax)
    ],
    extra_sympy_mappings={"inv": lambda x: 1 / x},
    # ^ Define operator for SymPy as well
    loss="loss(x, y) = (x - y)^2",
    # ^ Custom loss function (julia syntax)
)
model.fit(x,y)
```
 the following error appeared:
```py
RuntimeError: <PyCall.jlwrap (in a Julia function called from Python)
JULIA: tried to read a stream that is not readable
Stacktrace:
 [1] watch_stream(stream::Base.PipeEndpoint)
   @ SymbolicRegression.SearchUtilsModule ~/.julia/packages/SymbolicRegression/37l4B/src/SearchUtils.jl:79
 [2] _EquationSearch(::SymbolicRegression.CoreModule.ProgramConstantsModule.SRThreaded, datasets::Vector{SymbolicRegression.CoreModule.DatasetModule.Dataset{Float32}}; niterations::Int64, options::Options{typeof(loss), Int64, 0.86, 10}, numprocs::Nothing, procs::Nothing, addprocs_function::Nothing, runtests::Bool, saved_state::Nothing)
   @ SymbolicRegression ~/.julia/packages/SymbolicRegression/37l4B/src/SymbolicRegression.jl:418
 [3] EquationSearch(datasets::Vector{SymbolicRegression.CoreModule.DatasetModule.Dataset{Float32}}; niterations::Int64, options::Options{typeof(loss), Int64, 0.86, 10}, parallelism::String, numprocs::Nothing, procs::Nothing, addprocs_function::Nothing, runtests::Bool, saved_state::Nothing)
   @ SymbolicRegression ~/.julia/packages/SymbolicRegression/37l4B/src/SymbolicRegression.jl:383
 [4] EquationSearch(X::Matrix{Float32}, y::Matrix{Float32}; niterations::Int64, weights::Nothing, varMap::Vector{String}, options::Options{typeof(loss), Int64, 0.86, 10}, parallelism::String, numprocs::Nothing, procs::Nothing, addprocs_function::Nothing, runtests::Bool, saved_state::Nothing, multithreaded::Nothing)
   @ SymbolicRegression ~/.julia/packages/SymbolicRegression/37l4B/src/SymbolicRegression.jl:320
 [5] invokelatest(::Any, ::Any, ::Vararg{Any}; kwargs::Base.Pairs{Symbol, Any, NTuple{8, Symbol}, NamedTuple{(:weights, :niterations, :varMap, :options, :numprocs, :parallelism, :saved_state, :addprocs_function), Tuple{Nothing, Int64, Vector{String}, Options{typeof(loss), Int64, 0.86, 10}, Nothing, String, Nothing, Nothing}}})
   @ Base ./essentials.jl:731
 [6] _pyjlwrap_call(f::Function, args_::Ptr{PyCall.PyObject_struct}, kw_::Ptr{PyCall.PyObject_struct})
   @ PyCall ~/.julia/packages/PyCall/twYvK/src/callback.jl:32
 [7] pyjlwrap_call(self_::Ptr{PyCall.PyObject_struct}, args_::Ptr{PyCall.PyObject_struct}, kw_::Ptr{PyCall.PyObject_struct})
   @ PyCall ~/.julia/packages/PyCall/twYvK/src/callback.jl:44>
```
One reason might be that my Julia version is currently 1.8.5 instead of 0.6.0 (which isn't supported anymore, when I tried to downgrade the current version), or it could be another issue.
Is there any way to fix the issue?
When I tried running the code with an older Julia version (1.0), it said I had to update Julia first.
Thank you.

**Comment by MilesCranmer** [comment, marked as answer]
Any Julia version greater or equal to 1.6 should work fine (I am personally using 1.9.0-beta2). It could be that an older Julia version is being called instead? Are the older versions of Julia deleted from your machine?

Also, please provide system details (OS, Python version, whether you are using it in IPython/Jupyter/terminal).

**Reply by quasars100** [reply]
Since the older Julia version has been deleted and uninstalled (manually deleted the directories, called Julia to make sure it was no longer there, then re-installed the latest version), that's unlikely, but I will check the ~/bin directory again.
I am running Pysr in Python 3.9.0 (Jupyter notebook kernel, numpy 1.23.1, simply 1.6.2, scikit_learn 1.0.2), with MacOS Mojave (version 10.14.6).
Thank you.

**Reply by MilesCranmer** [reply]
Everything sounds fine to me. Are you using pip or conda?

Could you try this at the beginning of your notebook? It might be something with Julia not playing well with Jupyter notebook streams.
```python
from julia import Main
from julia.tools import redirect_output_streams

redirect_output_streams()
```
if it gives you an error, then run `from julia import Julia; julia = Julia(compiled_modules=False)` before that code (only required if you are using a static Python binary, like from `conda`).

**Reply by quasars100** [reply]
That worked (with "from julia import Julia; julia = Julia(compiled_modules=False" in front), thank you so much!
(Not sure if this will be helpful for others in the future, but also needed "import pysr; pysr.install()" in the notebook).
Thank you for all your help.

**Reply by MilesCranmer** [reply]
Thanks for letting me know. Could you provide more details on your environment? Is this conda python or system python or homebrew python? Are you running Jupyter in regular notebooks/JupyterLab, or through VSCode?

This is probably an actual bug, so I appreciate any details! Especially if I can locally reproduce it, then I can have PySR automatically set up this stream fix, rather than needing people to run it manually.

**Reply by quasars100** [reply]
Of course - my environment is system python (version 3.9.0), regular Jupyter notebook (no virtual environment setup).  The Julia version I am running is 1.8.5, installed using homebrew (brew install --cask julia); to make it compatible with python, I ran "import julia; julia.install()" in python3 (using terminal).  Then, using 

import pysr
pysr.install()

from julia import Julia; julia = Julia(compiled_modules=False)
from julia import Main
from julia.tools import redirect_output_streams

redirect_output_streams()

Pysr was able to run model.fit(x,y).  Hope this helps!

**Reply by MilesCranmer** [reply]
Thanks! I think you shouldn’t need the `from julia import Julia; julia = Julia(compiled_modules=False)` - is that correct?

**Reply by quasars100** [reply]
That's correct (it works with and without that line),

**Reply by MilesCranmer** [reply]
Thanks. So PySR 0.11.12, once it's up, will fix this issue, so you won't need to manually tweak the startup. The fix was this: https://github.com/MilesCranmer/SymbolicRegression.jl/pull/169

**Reply by quasars100** [reply]
Thank you!
